Upgrades202409 part2

VirginiaDooley opened the pull request:
Ref https://app.asana.com/0/1204880927741389/1208303037023014/f
Closes: 
https://democracy-club-gp.sentry.io/share/issue/4a5e72f0cdf341ae943d0596289f97bf/
https://democracy-club-gp.sentry.io/share/issue/41a20ab06e774a9b8aeccce5f39a81d2/
https://democracy-club-gp.sentry.io/issues/5899045389/?environment=staging&project=163593&query=is%3Aunresolved+issue.priority%3A%5Bhigh%2C+medium%5D&referrer=issue-stream&statsPeriod=90d&stream_index=0

co-authoured with @symroe 

BEFORE MERGING: 
- [ ] Remove staging branch
- [x] fix widget error https://democracy-club-gp.sentry.io/issues/5934646598/?project=163593&query=is%3Aunresolved%20issue.priority%3A%5Bhigh%2C%20medium%5D&referrer=issue-stream&statsPeriod=90d&stream_index=0

This PR implements upgrades to Python 3.12/Django 4.2.15/Postgres 16 as well as implements fixes to the [staging deploy](https://stage.electionleaflets.org/) that were not caught in the previous PR. 

The changes @symroe made to StaticStorage in https://github.com/DemocracyClub/electionleaflets/pull/255/commits/4cbc24c482ede6fbd22ba833d78baee640abb8c3 are from DC utils. Once these changes are merged, we should add `dc utils` and rip that code out of this project.

In addition to these changes, the DNS for [stage.electionleaflets.org (http://stage.electionleaflets.org/) wasn't hooked up. The prod EL account didn't have a stage NS set up. Sym added a NS record to electionleaflets.org pointing at the hosted zone in the stage EL account (that is set up).

Up next: 

- [ ] Review this list https://github.com/DemocracyClub/electionleaflets/security/dependabot
- [ ] Upload multiple files when JS is disabled https://docs.djangoproject.com/en/4.2/topics/http/file-uploads/#uploading-multiple-files
- [ ] Remove dead and unused code
- [ ] Install dc_utils to handle static files storage settings
- [ ] Adjust design-system for EL

chris48s commented on .circleci/config.yml:
```diff
@@ -75,13 +81,14 @@ jobs:
 
   sam_build:
     docker:
-      - image: amazon/aws-sam-cli-build-image-python3.8:latest
+      - image: public.ecr.aws/sam/build-python3.12:1.124.0-20240911233354
```
This seems like a really specific image to pin to.
On all the other projects where we are using the SAM images, we're targetting
`public.ecr.aws/sam/build-python3.12:latest`
any particular reason for this?

VirginiaDooley replied:
No, this was the latest version I found, but can loosen `public.ecr.aws/sam/build-python3.12:latest`

VirginiaDooley replied:
Fixed with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/9e32842c4153b22a8ed38a4d335698cef83ed8c9.

chris48s commented on .circleci/config.yml:
```diff
@@ -117,9 +125,27 @@ jobs:
+            pip install --upgrade pipenv==2024.0.1 pip --quiet --no-input
+      - run: pipenv sync --dev
+      - run: pip install -r deploy-requirements.txt
+      - run:
+          name: Check requirements
+          command: |
+            which make
+            pipenv run which make
+            echo $PATH
+            pipenv run echo $PATH
+            make --version
+            pipenv --venv
+            python --version
+            pipenv run python -m django --version
+            echo $DJANGO_SETTINGS_MODULE
+            pipenv run pip freeze
```
This whole section feels like debug code that could be removed.

chris48s commented on .circleci/config.yml:
```diff
@@ -222,7 +250,7 @@ workflows:
             - sam_build                       # deploy needs .aws-sam/build/
           context: [ deployment-staging-election-leaflets ]
-          filters: { branches: { only: [ master ] } }
+          filters: { branches: { only: [ master, upgrades202409-part2 ] } }
```
We should remove this branch filter before merging this PR

VirginiaDooley replied:
I've added a note in the PR description as a reminder.

VirginiaDooley replied:
This is done with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/26b3ee2deeeffdc3e1541fe1634a79fa6f8f6c5a

chris48s commented on Pipfile:
```diff
@@ -8,16 +8,15 @@ verify_ssl = true
 [packages]
 boto3 = "*"
-aws-wsgi = {ref = "a5b800373c326e75f705d400c48f7df301409724", git = "https://github.com/DemocracyClub/awsgi.git"}
+aws-wsgi = {git = "https://github.com/DemocracyClub/awsgi.git"}
```
There are 3 VCS dependencies in this file
- aws-wsgi
- dc_design_system
- django-uk-political-parties
I'm aware that they're pinned to a commit hash in Pipfile.lock but I think it would make it easier to understand what version we are _intending_ to be on if we pin all 3 to a ref (tag or commit) in Pipfile too.

VirginiaDooley replied:
Fixed with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/1bb47cdcc47a4138d736b166d4e51eb08a2372b5. I haven't updated the `deploy-requirements.txt` file until you confirm what's needed there.

chris48s commented on thumbs/Pipfile:
```diff
@@ -6,11 +6,11 @@ verify_ssl = true
 
 [requires]
-python_version = "3.7"
+python_version = "3.12.5"
```
This is going to be an annoyingly specific pin. Can we make this just 
python_version = "3.12"

VirginiaDooley replied:
Fixed with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/fdb568fc537adc00a94dc24c9a155a9aa859d119.

chris48s commented on Pipfile:
```diff
@@ -80,11 +78,13 @@ django-filter = "*"
 flask = "==2.3.2"
 certifi = "*"
+Markdown = "*"
 
```
can you add a 

```
[requires]
python_version = "3.12"
```

in the main `Pipfile` so that when I create an env locally it creates the env with the right python version without me having to specify it manually.

VirginiaDooley replied:
Fixed with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/20fa7d967e09a077f6e4cf732c86182879def80d.

chris48s commented on README.md:
```diff
@@ -14,8 +14,8 @@ In order to speed up this port, we are running the first version using the origi
-- Python 3.8.x
-- Node.JS 7+
+- Python 3.12.x
+- Node.js
```
Can we update the `engines` declaration in package.json to

```
"engines": {
  "node": ">=18"
}
```

VirginiaDooley replied:
Fixed with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/d57846a4c2f7cac802d1a2d77149b5c67d8d73c9

chris48s replied:
https://github.com/DemocracyClub/electionleaflets/pull/255#discussion_r1775428093

chris48s commented:
I've left some comments to follow up on here.

I think there is a further issue though, which is that I think we're probably duplicating requirements in the main Pipfile and deploy-requirements.txt and just installing everything everywhere rather than what we need where we need it.

I reckon to give proper actionable feedback on this I need to do a bit of a spike so I'm going to do that first but you might as well get on with the easy comments while I am doing that.

VirginiaDooley commented:
> Having a look over staging there's a couple of things I noticed.
> 
> 1. Dark text on dark background in the footer
> 
> ![GIFrecord_2024-09-25_093150](https://private-user-images.githubusercontent.com/6025893/370711904-56881485-842d-4602-8741-f757d7039296.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNjkwNDksIm5iZiI6MTcyNzI2ODc0OSwicGF0aCI6Ii82MDI1ODkzLzM3MDcxMTkwNC01Njg4MTQ4NS04NDJkLTQ2MDItODc0MS1mNzU3ZDcwMzkyOTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTI1MjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVkMTdmMDljMzhmYTVhN2FlZDdlYzc0MDk3MWU2NmE4YTk3MTBmNmMxY2I4YmZlNzNlYWUzZDJmN2FjZTUzMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EQOJ4V_ODJJ3EpDPvZIUfLLn7-Z7RrwmE_xJKnjHdJM) [ ![GIFrecord_2024-09-25_093150](https://private-user-images.githubusercontent.com/6025893/370711904-56881485-842d-4602-8741-f757d7039296.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNjkwNDksIm5iZiI6MTcyNzI2ODc0OSwicGF0aCI6Ii82MDI1ODkzLzM3MDcxMTkwNC01Njg4MTQ4NS04NDJkLTQ2MDItODc0MS1mNzU3ZDcwMzkyOTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTI1MjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVkMTdmMDljMzhmYTVhN2FlZDdlYzc0MDk3MWU2NmE4YTk3MTBmNmMxY2I4YmZlNzNlYWUzZDJmN2FjZTUzMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EQOJ4V_ODJJ3EpDPvZIUfLLn7-Z7RrwmE_xJKnjHdJM) ](https://private-user-images.githubusercontent.com/6025893/370711904-56881485-842d-4602-8741-f757d7039296.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNjkwNDksIm5iZiI6MTcyNzI2ODc0OSwicGF0aCI6Ii82MDI1ODkzLzM3MDcxMTkwNC01Njg4MTQ4NS04NDJkLTQ2MDItODc0MS1mNzU3ZDcwMzkyOTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTI1MjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVkMTdmMDljMzhmYTVhN2FlZDdlYzc0MDk3MWU2NmE4YTk3MTBmNmMxY2I4YmZlNzNlYWUzZDJmN2FjZTUzMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EQOJ4V_ODJJ3EpDPvZIUfLLn7-Z7RrwmE_xJKnjHdJM) [ ](https://private-user-images.githubusercontent.com/6025893/370711904-56881485-842d-4602-8741-f757d7039296.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNjkwNDksIm5iZiI6MTcyNzI2ODc0OSwicGF0aCI6Ii82MDI1ODkzLzM3MDcxMTkwNC01Njg4MTQ4NS04NDJkLTQ2MDItODc0MS1mNzU3ZDcwMzkyOTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTI1MjI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVkMTdmMDljMzhmYTVhN2FlZDdlYzc0MDk3MWU2NmE4YTk3MTBmNmMxY2I4YmZlNzNlYWUzZDJmN2FjZTUzMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EQOJ4V_ODJJ3EpDPvZIUfLLn7-Z7RrwmE_xJKnjHdJM)
> 
> This isn't the same on production. I assume we've changed something in this PR that breaks this?
> 
> 2. If I look at for example https://stage.electionleaflets.org/leaflets/27/ it is linked to a party, but the party doesn't exist. In fact there's no parties at all https://stage.electionleaflets.org/parties/ I don't mind if this is just because the data on staging is a bit broken, but I'd like to have confidence that this isn't a problem we've introduced in this PR.

This might be the result of removing the ref from the design system package, I'll look into it.

chris48s commented:
OK. Here's my suggested diff for the packages:

```diff
diff --git a/.circleci/config.yml b/.circleci/config.yml
index e58e7ac..1ba37d7 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -25,7 +25,6 @@ jobs:
           name: Set up virtual environment
           command: |
             pipenv sync --dev
-      - run: pip install -r deploy-requirements.txt
       - run: make lambda-layers/DependenciesLayer/requirements.txt
       - run: pipenv run collectstatic
       - run: npm install
diff --git a/deploy-requirements.txt b/deploy-requirements.txt
index 72869f1..37b7452 100644
--- a/deploy-requirements.txt
+++ b/deploy-requirements.txt
@@ -1,6 +1,2 @@
 aws-sam-cli==1.115.0
 urllib3==2.2.3
-Django==4.2.15
-git+https://github.com/DemocracyClub/design-system.git
-git+https://github.com/DemocracyClub/awsgi.git
-sentry_sdk==2.14.0
```

Basically:

- We don't need to install `deploy-requirements.txt` in the `install_and_update_dependencies` because we're not trying to deploy anything yet. We only need that once we get to the build/deploy jobs
- None of those other packages are needed in the deploy step

I guess with the extra packages in `deploy-requirements.txt` were hitting errors in the deploy step so you started adding more packages to `deploy-requirements.txt`, then added `pipenv sync --dev` which fixed them and didn't take the extra packages back out of `deploy-requirements.txt` ?

Ta

VirginiaDooley commented:
> OK. Had a quick look at it. Here's my suggested diff for the packages:
> 
> ```diff
> diff --git a/.circleci/config.yml b/.circleci/config.yml
> index e58e7ac..1ba37d7 100644
> --- a/.circleci/config.yml
> +++ b/.circleci/config.yml
> @@ -25,7 +25,6 @@ jobs:
>            name: Set up virtual environment
>            command: |
>              pipenv sync --dev
> -      - run: pip install -r deploy-requirements.txt
>        - run: make lambda-layers/DependenciesLayer/requirements.txt
>        - run: pipenv run collectstatic
>        - run: npm install
> diff --git a/deploy-requirements.txt b/deploy-requirements.txt
> index 72869f1..37b7452 100644
> --- a/deploy-requirements.txt
> +++ b/deploy-requirements.txt
> @@ -1,6 +1,2 @@
>  aws-sam-cli==1.115.0
>  urllib3==2.2.3
> -Django==4.2.15
> -git+https://github.com/DemocracyClub/design-system.git
> -git+https://github.com/DemocracyClub/awsgi.git
> -sentry_sdk==2.14.0
> ```
> 
> Basically:
> 
> * We don't need to install `deploy-requirements.txt` in the `install_and_update_dependencies` because we're not trying to deploy anything yet. We only need that once we get to the build/deploy jobs
> * None of those other packages are needed in the deploy step
> 
> I guess with the extra packages in `deploy-requirements.txt` were hitting errors in the deploy step so you started adding more packages to `deploy-requirements.txt`, then added `pipenv sync --dev` which fixed them and didn't take the extra packages back out of `deploy-requirements.txt` ?
> 
> Ta

Exactly. That's now fixed with https://github.com/DemocracyClub/electionleaflets/pull/255/commits/fb011d466d9f17983a1ab6f71c719e93e231c6d7.

chris48s commented on package-lock.json:
```diff
@@ -16,7 +16,7 @@
         "gulp-concat": "^2.6.1"
       },
       "engines": {
-        "node": "> 7"
+        "node": ">=18"
```
We're only changing the lockfile in this commit. Did you forget to add `package.json` to the commit here or did you manually edit `package-lock.json`?

In general the process should be:
- Edit `package.json`
- Run `npm install` to rebuild `package-lock.json`

VirginiaDooley replied:
Fixed! https://github.com/DemocracyClub/electionleaflets/pull/255/commits/15bf573c409f9b6d71d98d7591a68eed585f5095

chris48s commented:
> UPDATE: The version of design-system used on master is the very first version. Since then, changes have been made to the design system that have not accounted for the child <p> elements which (I think) only appear in EL. This will require adding that flexibility to the design-system. Shall I put a temp fix in here?

Yeah I think lets just work round it here if we can. Can we bing the footer more into line with what we're doing on other applications or bodge it by putting it in a `<div class="ds-copyright">` (even though that's not really the right class name) or something?

I reckon just binning that text from the footer is probably better than leaving the dark text on dark background.

Did you have a chance to think about the missing parties on staging at all?

VirginiaDooley commented:
> > UPDATE: The version of design-system used on master is the very first version. Since then, changes have been made to the design system that have not accounted for the child  elements which (I think) only appear in EL. This will require adding that flexibility to the design-system. Shall I put a temp fix in here?
> 
> Yeah I think lets just work round it here if we can. Can we bing the footer more into line with what we're doing on other applications or bodge it by putting it in a `<div class="ds-copyright">` (even though that's not really the right class name) or something?
> 
> I reckon just binning that text from the footer is probably better than leaving the dark text on dark background.
I think this is the best solution to remove that text and add a link to DC website similar to other sites

<img width="360" alt="Screenshot 2024-09-25 at 4 49 36 PM" src="https://github.com/user-attachments/assets/6d7828b4-00d7-4e85-ac1e-e1d34ef6dd94">

> Did you have a chance to think about the missing parties on staging at all?
 Will look into this now.

VirginiaDooley commented on electionleaflets/apps/leaflets/views.py:
```diff
@@ -222,21 +222,33 @@ def done(self, form_list, **kwargs):
                     new_file_name_raw = f"raw_{new_file_name}"
-                    moved_file_raw = bucket.Object(new_file_name_raw)
-                    moved_file_raw.copy(copy_source)
 
```
@chris48s This change fixes local uploads, although thumbnails still need attention.

symroe commented on electionleaflets/apps/leaflets/forms.py:
```diff
@@ -33,7 +33,7 @@ def __init__(self, **kwargs):
             self.fields["image"] = S3UploadedImageField(
                 widget=forms.ClearableFileInput(
-                    attrs={"multiple": True, "accept": "image/*",}
+                    attrs={"accept": "image/*",}
```
Will this remove the ability to upload more than one leaflet image at a time? Can you check?

VirginiaDooley replied:
Multiple uploads still possible. 
<img width="725" alt="Screenshot 2024-09-30 at 9 52 05 AM" src="https://github.com/user-attachments/assets/5010e5b7-5d22-4b6e-8b14-e8ac6dcbac45">

chris48s replied:
Throwing an exception on this was a change made in the release of Django 3.2.19
https://docs.djangoproject.com/en/4.2/releases/3.2.19/
so we can't just pass `"multiple": True` here on the version of Django we're moving to.

This seems to be the recommended way to allow a single form field to submit >1 file now
https://docs.djangoproject.com/en/4.2/topics/http/file-uploads/#uploading-multiple-files
but I'm not 100% clear on whether leaflets actually relies on that.

I'm unclear from the docs whether this code ever did what we thought it did tbh.
